asm/kanread.s

```diff
@@ -0,0 +1,206 @@
+.include "macro.inc"
+
+# assembler directives
+.set noat      # allow manual use of $at
+.set noreorder # don't insert nops after branches
+.set gp=64     # allow use of 64-bit general purpose registers
+
+.section .text
+
+.balign 16
+
+# s32 Kanji_OffsetFromShiftJIS(s32 shiftJISCodepoint) // Originally LeoGetKAdr
+# 
+# Returns the offset of the glyph texture data in the file `kanji` corresponding
+# to a given 2-byte Shift-JIS codepoint. No range validity check is carried out.
+#
+# A nice Shift-JIS codepoint table: https://uic.io/en/charset/show/shift_jis/
+# The file `kanji` contains the 'Level 1' kanji (0x889F-0x9872), and a reworked
+# version of the non-kanji section that includes extra English and Hylian glyphs.
+#
+# @param shiftJISCodepoint Codepoint of glyph.
+# return s32 offset into `kanji` file.
+BEGIN Kanji_OffsetFromShiftJIS
+    # Characters with codepoints >= 0x8800 are kanji. Arrangement is regular, 
+    # so convert index directly.
+    li      $at, 0x8800
+    slt     $at, $a0, $at
+    bnez    $at, .not_kanji
+     # 0xBC is the size of one block of glyphs in the `kanji` file:
+     # 0x100 possible codepoints with the same byte1
+     # - 0x40 unused at beginning 
+     # - 1 unused at 0x7F 
+     # - 3 unused at 0xFD, 0xFE, 0xFF
+     li     $a2, 0xBC
+    # Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x88
+    multu   $a2, $a1
+    # Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xff
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    # 0xXX7F is always empty and elided in the file, so decrement
+    bnezl   $at, .kanji_lower_block
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.kanji_lower_block:
+    addi    $a3, $a3, 0x30a
+    add     $a3, $a3, $a2
+    jr      $ra
+     sll    $v0, $a3, 7
+     # returns (0x30A + (adjusted byte2) + (adjusted byte1) * 0xBC) * FONT_CHAR_TEX_SIZE
+
+# Non-kanji are arranged with irregular gaps, use the lookup table.
+.not_kanji:
+    # Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x81
+    multu   $a2, $a1
+    # Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xff
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
+    mflo    $a2
+    # 0xXX7F is always empty and elided in the file, so decrement
+    bnezl   $at, .not_kanji_lower_block
```

Comment:
I initially misread this label as "not kanji-lower-block" instead of "not-kanji lower-block". But I don't have a suggestion to prevent that confusion

Reply:
Yeah, there's not really anything that ties this section together apart from the glyphs not being kanji. I suppose `non_kanji` or `nonkanji` would be an option, but I don't think it's really an improvement.

Reply:
maybe `not_kanji__lower_block` (notice the double underscore in the middle)